safep/TI.py

```diff
@@ -49,8 +65,53 @@ def plot_TI(cumulative, perWindow, width=8, height=4, PDFtype='KDE', hystLim=(-1
     
     return fig, [cumAx,eachAx] 
 
-def make_harmonicWall(FC=10, targetFC=0, targetFE=1, upperWalls=1, schedule=None, numSteps=1000, targetEQ=500, name='HW', lowerWalls=None):
-    HW = {'name':name, 'targetFC':targetFC, 'targetFE':targetFE, 'FC':FC, 'upperWalls':upperWalls, 'schedule':schedule, 'numSteps':numSteps, 'targetEQ':targetEQ, 'lowerWalls':lowerWalls}
+def make_harmonicWall(FC=10, targetFC=0, targetFE=1, upperWalls=1, schedule=None, numSteps=1000, targetEQ=500, name='HW', lowerWalls=None, lambdaExp=1., decoupling=None):
+    '''
+    This method should rarely be used, as long as a log file exists that can be parsed by parse_Colvars_log
+    and the result fed to make_harmonicWall_from_Colvars
+    '''
+
+    # Heuristic to set the default value of the decoupling parameter
+    if decoupling is None:
+        decoupling = (targetFC == 0 and FC != 0)
+
+    HW = {'name':name, 'targetForceConstant':targetFC, 'lambdaExponent':targetFE, 'forceConstant':FC, 'upperWalls':upperWalls, 'schedule':schedule,
+           'targetNumSteps':numSteps, 'targetEquilSteps':targetEQ, 'lowerWalls':lowerWalls, 'lambdaExp':lambdaExp, 'decoupling':decoupling}
+
+    return HW
+
+def make_harmonicWall_from_Colvars(w):
+    '''
+    Input: a restraint configuration as a dict produced by parse_Colvars_log
+    '''
+
+    if (w['key'] != 'harmonicwalls'):
+        k = w['key']
+        print(f'Error: bias is not a harmonic wall (key: {k})')
+        return None
+
+    CVs = w['colvars'].strip("{}").replace(",", "").split()
+    if len(CVs) != 1:
+        print(f'Error: bias does not act on exactly one colvar (cvs: {CVs})')
+        return None
+
+    HW = {  'name': w['name'],
+            'colvar': CVs[0],
+            'forceConstant': float(w['forceConstant']),
+            'targetForceConstant': float(w['targetForceConstant']),
+            'lowerWalls': float(w['lowerWalls'].strip('{ }')),
+            'upperWalls': float(w['upperWalls'].strip('{ }')),
+            'targetNumSteps': int(w['targetNumSteps']),
+            'targetNumStages': int(w['targetNumStages']),
+            'targetEquilSteps': int(w['targetEquilSteps']),
+            'decoupling': w['decoupling'] in ['on', 'true', 'yes']} # interpret as Boolean
+    # Support legacy keyword
+    if 'targetForceExponent' in w:
+        HW['lambdaExponent'] = int(w['targetForceExponent'])
+    if 'lambdaExponent' in w:
+        HW['lambdaExponent'] = int(w['lambdaExponent'])
+    cleaned = w['lambdaSchedule'].strip("{}").replace(",", "")

```

Comment:
I don't quite follow these two lines. The sample log has the line:
`lambdaSchedule =  [default]`
So would these lines set HW['schedule'] to just an empty list?

Reply:
That appears to be what it's doing in the sample notebook. Is that expected? Do we even need this variable?

Reply:
We do! The integration depends on the lambda schedule, see this in the notebook (for now):
```
if DBC_rest['schedule'] == []:
    schedule = np.linspace(0, 1, nLambdas)
else:
    schedule = DBC_rest['schedule']
```
I've just moved this first part to the `make_harmonicWall_from_Colvars` function.
```
dataLs = np.round([schedule[i] for i in Ls], 3)
dataTI.loc[:,'L'] = dataLs
```
Now I'm wondering if we should get rid of this rounding of lambda values, I don't think it helps here and it can break integration for very small steps.